### PR TITLE
global variable ver 1 pattern match

### PR DESCRIPTION
**Pull request by ChengdaXu**

### CODE REVIEW

**Comment by izgzhen** [comment]
Thanks! Please add the testing code I mentioned last time as well.

**Comment by ChengdaXu** [comment]
> Thanks! Please add the testing code I mentioned last time as well.

Done.

**Comment by izgzhen** [comment]
```
[ERROR] Tests run: 7, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 10.467 s <<< FAILURE! - in edu.washington.cs.seguard.AppTest
[ERROR] testJS2(edu.washington.cs.seguard.AppTest)  Time elapsed: 2.95 s  <<< FAILURE!
java.lang.AssertionError: 
```

CI failed.

**Comment by ChengdaXu** [comment]
See details in issue#2 https://github.com/semantic-graph/seguard-java/issues/2

File: src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala [hunk 1]
```diff
@@ -131,18 +133,28 @@ object JSFlowGraph {
       }
     }
   }
+  def manOf[T: Manifest](t: T): Manifest[T] = manifest[T]
```
Comment (izgzhen) [inline src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala]: I didn't see it used anywhere

File: src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala [hunk 2]
```diff
@@ -192,7 +202,19 @@ object JSFlowGraph {
           }
 
           // DefUse based analysis
-          val u = currentNode.get
+          var u = currentNode.get
+          val namespace = bb.getNode.toString
+          if (!globalVarMap.contains(namespace)) {
+            globalVarMap.put(namespace, new HashMap());
```
Comment (izgzhen) [inline src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala]: what is `namespace`?

**Comment by izgzhen** [comment]
Thanks! Just some minor issue as above plus info in the email.

File: src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala (hunk 1 above)
Comment (ChengdaXu) [reply src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala]: I deleted it.

File: src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala (hunk 2 above)
Comment (ChengdaXu) [reply src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala]: Here namespace refers to the name of the function where these variables are defined. For example, in different call graph node (or, in other words, function), v11 may refer to different variables.

Comment (izgzhen) [reply src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala]: Okay, I see. Please add this comment to code.

**Comment by izgzhen** [comment]
Also, why isn't the edges file not updated?

File: src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala [hunk 3]
```diff
@@ -265,3 +290,38 @@ object JSFlowGraph {
     }
   }
 }
+//          // DefUse based analysis
```
Comment (izgzhen) [inline src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala]: Why are you commenting out a large block of code?

File: src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala [hunk 4]
```diff
@@ -235,7 +247,20 @@ object JSFlowGraph {
           abstractInstruction(node.getNode.getDU, symTable, instruction) match {
             case Some(u) => {
               // DefUse based analysis
-              dot.drawNode(u, NodeType.STMT)
+              var u_complete = u
```
Comment (izgzhen) [inline src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala]: What does `u_complete` mean?

File: src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala (hunk 3 above)
Comment (ChengdaXu) [reply src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala]: Sorry I forgot to remove it. I copy the codes as reference when merging with your codes. It was removed now.

File: src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala (hunk 4 above)
Comment (ChengdaXu) [reply src/main/scala/edu/washington/cs/seguard/js/JSFlowGraph.scala]: It means the complete name of u. For example, when u is [get]env, and n_complete would be process[env]

**Pull request merged**